File: surrealml/clients/typescript/scripts/update_version.ts
#!/usr/bin/env node
import { promises as fs } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";


const SEMVER_RE = /^\d+\.\d+\.\d+$/;
const __filename = fileURLToPath(import.meta.url);
const __dirname  = dirname(__filename);
const CONFIG_JSON_PATH = join(__dirname, "..", "config.json");


function validateSemver(v: string, label: string): string {
	if (!SEMVER_RE.test(v)) {
		console.error(`Invalid ${label} '${v}': must be in format X.Y.Z`);
		process.exit(1);
	}
	return v;
}

async function main() {
    const args = process.argv.slice(2);

    // Expect: <npm_package_version> [dynamic_lib_version]
    if (args.length < 1 || args.length > 2 || !args[0]) {
		console.error("Usage: bump-version <npm_package_version> [dynamic_lib_version]");
		process.exit(1);
    }

    const newNpm = validateSemver(args[0], "npm_package_version");
    const newDyn = args[1] ? validateSemver(args[1], "dynamic_lib_version") : undefined;

    // load config.json
    let cfg: Record<string, any>;
    try {
		const raw = await fs.readFile(CONFIG_JSON_PATH, "utf-8");
		cfg = JSON.parse(raw);

    } catch (err: any) {
		console.error(`Error reading '${CONFIG_JSON_PATH}': ${err.message}`);
		process.exit(1);
    }

    // always update npm_package_version
    cfg.npm_package_version = newNpm;

    // optionally update dynamic_lib_version
    if (newDyn) cfg.dynamic_lib_version = newDyn;

    // write back
    try {
		const out = JSON.stringify(cfg, null, 4) + "\n";
		await fs.writeFile(CONFIG_JSON_PATH, out, "utf-8");
    } catch (err: any) {
		console.error(`Error writing '${CONFIG_JSON_PATH}': ${err.message}`);
		process.exit(1);
    }

    console.log("✅  config.json updated");
}

main();

File: surrealml/clients/python/scripts/update_version.py
#!/usr/bin/env python3
import json
import sys
import argparse
import re
from pathlib import Path


# ─── Constants ──────────────────────────────────────────────────────────────
SEMVER_RE = re.compile(r'^\d+\.\d+\.\d+$')
CONFIG_JSON_PATH = Path(__file__).parent.parent / "config.json"

# ─── Helpers ────────────────────────────────────────────────────────────────
def semver_type(value: str) -> str:
    if not SEMVER_RE.match(value):
        raise argparse.ArgumentTypeError(
            f"Invalid version '{value}': must be in format X.Y.Z"
        )
    return value

def parse_args():
    p = argparse.ArgumentParser(
        description="Bump package version (and optionally dynamic-lib version) "
                    "inside config.json")
    p.add_argument("new_version",            type=semver_type,
                   help="new python_package_version (X.Y.Z)")
    p.add_argument("dynamic_lib_version",    type=semver_type, nargs='?',
                   help="optional: new dynamic_lib_version (X.Y.Z)")
    return p.parse_args()

def load_config() -> dict:
    try:
        with CONFIG_JSON_PATH.open('r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        print(f"Error reading '{CONFIG_JSON_PATH}': {e}", file=sys.stderr)
        sys.exit(1)

def write_config(cfg: dict) -> None:
    try:
        with CONFIG_JSON_PATH.open('w', encoding='utf-8') as f:
            json.dump(cfg, f, indent=4)
            f.write('\n')
    except Exception as e:
        print(f"Error writing '{CONFIG_JSON_PATH}': {e}", file=sys.stderr)
        sys.exit(1)

# ─── Main ───────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    args   = parse_args()
    cfg    = load_config()

    cfg["python_package_version"] = args.new_version
    if args.dynamic_lib_version is not None:
        cfg["dynamic_lib_version"] = args.dynamic_lib_version

    write_config(cfg)
    print("✅  config.json updated")
